rust/connlib/tunnel/src/control_protocol/client.rs

```diff
@@ -84,120 +91,88 @@ where
-                    .lock()
-                    .awaiting_connection
-                    .remove(&resource_id);
-            })
-        }));
+        let IceConnection {
+            ice_params,
+            ice_transport,
+            ice_candidate_rx,
+        } = new_ice_connection(&self.webrtc_api, relays).await?;
+        let preshared_key = self
+            .role_state
+            .lock()
+            .add_waiting_gateway(gateway_id, ice_candidate_rx);
+        self.peer_connections
+            .lock()
+            .insert(gateway_id, Arc::clone(&ice_transport));
 
-        let offer = peer_connection.create_offer(None).await?;
-        peer_connection.set_local_description(offer.clone()).await?;
+        set_connection_state_update(self, &ice_transport, gateway_id, resource_id);
 
         Ok(Request::NewConnection(RequestConnection {
             resource_id,
             gateway_id,
             client_preshared_key: Secret::new(Key(preshared_key.to_bytes())),
-            client_rtc_session_description: offer,
+            client_rtc_session_description: ice_params,
         }))
     }
 
+    fn new_tunnel(
+        &self,
+        resource_id: ResourceId,
+        gateway_id: GatewayId,
+        ice: Arc<RTCIceTransport>,
+    ) -> Result<()> {
+        let p_key = self
+            .role_state
+            .lock()
+            .gateway_preshared_keys
+            .get(&gateway_id)
+            .ok_or(Error::ControlProtocolError)?
+            .clone();
+        let index = self.next_index();
+
+        let peer_config = self
+            .role_state
+            .lock()
+            .create_peer_config_for_new_connection(resource_id, gateway_id, p_key)?;
+
+        let peer = Arc::new(Peer::new(
+            self.private_key.clone(),
+            index,
+            peer_config.clone(),
+            gateway_id,
+            None,
+            self.rate_limiter.clone(),
+        ));
+
+        let (peer_sender, peer_receiver) = tokio::sync::mpsc::channel(PEER_QUEUE_SIZE);
+
+        start_handlers(
+            Arc::clone(&self.device),
+            self.callbacks.clone(),
+            peer.clone(),
+            ice,
+            peer_receiver,
+        );
+
+        // Partial reads of peers_by_ip can be problematic in the very unlikely case of an expiration
+        // before inserting finishes.
+        insert_peers(
+            &mut self.peers_by_ip.write(),
+            &peer_config.ips,
+            ConnectedPeer {
+                inner: peer,
+                channel: peer_sender,
+            },
+        );
+
+        // Note: worst that can happen if peer_by_ip has been updated but the awaiting_* locks haven't is that we lose a reuse connection and it has to be retried.
+        // This is very unlikely and not an error and not having both locks at the same time greatly reduce the chance of a deadlock.
+        self.role_state
+            .lock()
+            .on_connection_success(&gateway_id, &resource_id);
```

Comment:
Nothing after `create_peer_config_for_new_connection` can fail so we should be able to inline this into that function, right?